lib/clowder-common-ruby/rails_config.rb

```diff
@@ -0,0 +1,175 @@
+
+      redis_config = {
+        url: redis_url,
+        password: inMemoryDb.dig('password')
+      }
+    end
+
+    # Nested map for Cloudwatch configuration
+    def configure_cloudwatch
+      cloudwatch = first_config.logging&.cloudwatch
+      
+      final_cloudwatch_config = {
+        region: cloudwatch&.region,
+        log_group: cloudwatch&.logGroup,
+        log_stream: Socket.gethostname,
+        type: first_config.logging&.type,
+        credentials: {
+          access_key_id: cloudwatch&.accessKeyId,
+          secret_access_key: cloudwatch&.secretAccessKey
+        }
+      }
+    end  
+
+    # Map for RBAC configuration 
+    def configure_rbac
+      rbac_config = first_config.dependency_endpoints.dig('rbac', 'service')
+
+      if first_config.tlsCAPath
+        rbac_host = "#{rbac_config.hostname}:#{rbac_config.tlsPort}"
+        rbac_scheme = 'https'
+      else
+        rbac_host = "#{rbac_config.hostname}:#{rbac_config.port}"
+        rbac_scheme = 'http'
+      end
+
+      final_rbac_config = {
+        host: rbac_host,
+        scheme: rbac_scheme
+      }
+    end
+
+    # Map for endpoint configuration 
+    def configure_endpoint(endpoint)
+      if first_config.tlsCAPath
+        scheme = 'https'
+        host = "#{endpoint.hostname}:#{endpoint.tlsPort}"
+        url = "#{scheme}://#{host}"
+      else
+        scheme = 'http'
+        host = "#{endpoint.hostname}:#{endpoint.port}"
+        url = "#{scheme}://#{host}"
+      end
+      
+      endpoint_config = {
+        host: host,
+        scheme: scheme,
+        url: url
+      }
+    end
+
+    # Nested map for Rails configuration
+    def self.to_hash
+      rails_config = new
+      first_config = rails_config.first_config
+
+      # Kafka
+      kafka_server_config = rails_config.configure_kafka
+
+      # Cloudwatch
+      cloudwatch_config = rails_config.configure_cloudwatch
+
+      # RBAC
+      rbac_config = rails_config.configure_rbac
+
+      # Redis
+      redis_config = rails_config.configure_redis
+
+      clowder_config = {
+        kafka: kafka_server_config,
+        logging: cloudwatch_config,
+        rbac: rbac_config,
+        redis: redis_config,
+        clowder_config_enabled: true,
+        prometheus_exporter_port: first_config&.metricsPort,
+        tlsCAPath: first_config.tlsCAPath
+      }
+
+      if first_config.dependency_endpoints.present?
+        first_config.dependency_endpoints.each do |name, endpoint|
+          key = "#{name}".underscore.to_sym
+          clowder_config[key] = rails_config.configure_endpoint(endpoint.dig('service'))
+        end
+      end
+      
+      if first_config.private_dependency_endpoints.present?
+        first_config.private_dependency_endpoints.each do |name, endpoint|
+          key = "#{name}".underscore.to_sym
+          clowder_config[key] = rails_config.configure_endpoint(endpoint.dig('service'))
+        end
+      end
```

Comment:
In general these two steps can be done during the hash creation above using `each_with_object` and you don't really need to adjust the hash afterwards:

```
def self.build_endpoints(endpoints)
  endpoints.&each_with_object({}) do |name, endpoint, obj| # not sure about the name-endpoint destruct here, maybe (name, endpoint), obj ... try it
    key = ...
    obj[key] = configure_endpoint(endpoint.dig('service'))
  end
end

def self.to_h
  clowder_config = {
     ...
     endpoints: build_endpoints(first_config&.dependency_endpoints),
     private_endpoints: build_endpoints(first_config&.private_dependency_endpoints)
     ...
  }
end
```